src/NHibernate/Type/CollectionType.cs

```diff
@@ -391,14 +392,36 @@ public virtual object InstantiateResult(object original)
 			// in which case we should never ever ever reset the dirty flag
 			// on the target because we simply do not know...
 			IPersistentCollection originalPc = original as IPersistentCollection;
-			IPersistentCollection resultPc = result as IPersistentCollection;
-			if(originalPc != null && resultPc!=null)
+			IPersistentCollection resultPc = target as IPersistentCollection;
+			if(resultPc!=null)
 			{
-				if (!originalPc.IsDirty)
+				if (elementsEqual || (originalPc != null && !originalPc.IsDirty))
 					resultPc.ClearDirty();
 			}
 
-			return result;
+			return target;
+		}
+
+		private static bool AreCollectionElementsEqual(IEnumerable originalEnumerator, IEnumerator targetEnumerator, IType elementType, ISessionImplementor session)
```

Comment:
@hazzik Thanks for your review.
Sure, I'll update PR soon.
(on the other hand makes current PR it work in >95% of the cases, it's only a problem if you on a detached set remove an element and put it back again...)